metaflow/metaflow_system_logger.py

```diff
@@ -0,0 +1,109 @@
+import sys
+from typing import Dict, Any
+
+
+class SystemLogger(object):
+    def __init__(self):
+        self._flow = None
+        self._environment = None
+        self._logger = None
+        self._flow_name = None
+
+    def __del__(self):
+        if self.flow_name == "not_a_real_flow":
+            self.logger.terminate()
+
+    @property
+    def environment(self):
+        from .plugins import ENVIRONMENTS
+        from .metaflow_config import DEFAULT_ENVIRONMENT
+        from .metaflow_environment import MetaflowEnvironment
+
+        if self._environment is None:
+            self._environment = [
+                e
+                for e in ENVIRONMENTS + [MetaflowEnvironment]
+                if e.TYPE == DEFAULT_ENVIRONMENT
+            ][0](self.flow)
+        return self._environment
+
+    @environment.setter
+    def environment(self, environment):
+        self._environment = environment
+
+    @property
+    def flow(self):
+        from metaflow.sidecar import DummyFlow
+
+        if self._flow is None:
+            self._flow = DummyFlow()
+            self.flow_name = "not_a_real_flow"
+        return self._flow
+
+    @flow.setter
+    def flow(self, flow):
+        self._flow = flow
+
+    @property
+    def flow_name(self):
+        return self._flow_name
```

Comment:
assumes that flow_name will not be called prior to flow which may not be correct. In general, I would have one init method that inits all 4 (flow, flow_name, environment and logger) and controlled by one single flag.

Reply:
This comment is still not addressed I think.